[vulkan][asan] Fix Invalid Memory ops

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42181 [vulkan] Permtute ops: transpose,view,slice,unsqueeze,select
* #42180 [vulkan][op] avg_pool2d
* #42179 [vulkan][op] add.Scalar, mul.Scalar
* #41434 [vulkan] cat op (concatenate)
* #41380 [vulkan] inplace add_, relu_
* #41379 [vulkan] max_pool2d
* **#41224 [vulkan][asan] Fix Invalid Memory ops**
* #41223 [vulkan] reshape op

Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

IvanKobzarev committed:
[vulkan][asan] Fix Invalid Memory ops

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"




[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanAten.cpp:
```diff
@@ -262,8 +262,7 @@ at::Tensor vulkan_convolution_prepacked(
   const bool hasBias = bias.has_value() && bias->defined();
-  const bool vulkanBias = (*bias).is_vulkan();
-  if (hasBias && vulkanBias) {
+  if (hasBias && (*bias).is_vulkan()) {
```
Alternatively you can write as
```
if (hasBias && bias->is_vulkan()) {
...
}
```
since the dereference operator is overloaded for c10::optional.  Not that it really matters but just maybe a tiny bit more idiomatic.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -259,12 +259,20 @@ VBuffer kernelNCHW_OCHW_repack_O4C4HWi4o4(
   VBuffer buffer{sizeAligned};
   if (data.has_value()) {
-    buffer.copy_from_host_to_device((void*)*data, size);
+    buffer.copy_from_host_to_device((void*)*data, dataSize);
```
Any pointer type can implicity cast to void *.  The cast should not be necessary.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -259,12 +259,20 @@ VBuffer kernelNCHW_OCHW_repack_O4C4HWi4o4(
     c10::optional<const float*> data,
-    const uint32_t size) {
+    const uint32_t dataSize,
+    const uint32_t bufferSize) {
```
I am not sure I understand why this function accepts bufferSize as a parameter as well when this function is the entity that ends up allocating the buffer itself a few lines below.  Is it because you want to put more data inside this buffer later on and want to make sure you have enough storage?

IvanKobzarev replied:
Yes, that is memory alignment. For example, if we want to bind in shader vec4[4],  our buffer should be 4 * 4 * 4 (float size). While we might have only 14 floats.

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"


Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"


Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"


Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"


Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"


Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"


Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][asan] Fix Invalid Memory ops"


Differential Revision: [D22754940](https://our.internmc.facebook.com/intern/diff/D22754940)

[ghstack-poisoned]
